File: src/workflows.ts
import { proxyActivities, startChild, uuid4 } from '@temporalio/workflow'
import * as activities from './activities'

const { createS3Bucket, deleteS3Object, deleteS3Bucket, getRelatedDocuments, invokePrompt, loadTestCases, getLatestDocumentProcessingId } = proxyActivities<typeof activities>({
  startToCloseTimeout: '1 minute',
  retry: {
    backoffCoefficient: 1,
    initialInterval: '3 seconds'
  }
})

const { collectDocuments, validateQueryResult, summarizeValidationResults } = proxyActivities<typeof activities>({
  startToCloseTimeout: '5 minute',
  retry: {
    backoffCoefficient: 1,
    initialInterval: '3 seconds'
  }
})

const { processDocuments } = proxyActivities<typeof activities>({
  startToCloseTimeout: '50 minute',
  retry: {
    backoffCoefficient: 1,
    initialInterval: '3 seconds'
  }
})

type Repository = {
  url: string
  branch: string
  path: string
  fileExtensions: string[]
}
type DocumentsProcessingWorkflowInput = {
  id: string
  repository: Repository
  failRate: number
}
type DocumentsProcessingWorkflowOutput = {
  tableName: string
}
export async function documentsProcessingWorkflow(input: DocumentsProcessingWorkflowInput): Promise<DocumentsProcessingWorkflowOutput> {
  const { id, repository, failRate } = input
  const { url, branch, path, fileExtensions } = repository

  await createS3Bucket({
    bucket: id,
    failRate
  })

  const { zipFileName } = await collectDocuments({
    workflowId: id,
    s3Bucket: id,
    gitRepoUrl: url,
    gitRepoBranch: branch,
    gitRepoDirectory: path,
    fileExtensions,
    failRate
  })

  const { tableName } = await processDocuments({
    workflowId: id,
    s3Bucket: id,
    zipFileName,
    failRate
  })

  await deleteS3Object({ bucket: id, key: zipFileName })

  await deleteS3Bucket({ bucket: id })

  return { tableName }
}

type QueryWorkflowInput = {
  query: string 
  failRate: number
}
type QueryWorkflowOutput = {
  conversationId: string
  response: string
}
export async function invokePromptWorkflow(input: QueryWorkflowInput): Promise<QueryWorkflowOutput> {
  const { query, failRate } = input 
  const latestDocumentProcessingId = await getLatestDocumentProcessingId()  
  const conversationId = `conversation-${uuid4()}`

  await createS3Bucket({
    bucket: conversationId,
    failRate
  })

  const { relatedDocumentsContentFileName } = await getRelatedDocuments({
    query,
    latestDocumentProcessingId,
    s3Bucket: conversationId,
    failRate
  })

  const { response } = await invokePrompt({
    query,
    s3Bucket: conversationId,
    relatedDocumentsContentFileName,
    failRate
  })

  return { conversationId, response }
}

type TestPromptsWorkflowInput = {
  latestDocumentProcessingId: string
  testName: string
}
type TestPromptsWorkflowOutput = {
  validationResults: {
    query: string
    answer: string
    score: number
    reason: string
  }[],
  summary: string
  averageScore: number
}
export async function testPromptsWorkflow(input: TestPromptsWorkflowInput): Promise<TestPromptsWorkflowOutput> {
  const testCases = await loadTestCases({
    testName: input.testName
  }) 
  console.log("Loaded test cases.") 
  const queries = Object.keys(testCases)


  const childWorkflowHandles = await Promise.all(
    queries.map((query) => {
      return startChild('invokePromptWorkflow', {
        taskQueue: 'invoke-prompt-queue',
        args: [{
          query,
          latestDocumentProcessingId: input.latestDocumentProcessingId,
        }]
      })
    }
  ));

  const childWorkflowResponses = await Promise.all(
    childWorkflowHandles.map(async (handle, index) => {
      const query = queries[index];
      const expectedResponse = testCases[query]
      const result = await handle.result();

      return {
        query,
        expectedResponse,
        actualResponse: result.response,
      }
    })
  );

  const validationResults = await Promise.all(
    childWorkflowResponses.map(async (input) => validateQueryResult(input))
  )

  const { summary, averageScore } = await summarizeValidationResults({ validationResults })

  return {
    validationResults,
    summary,
    averageScore
  }
}

File: src/workflows.go
package workflows

import (
	"fmt"
	"temporal-hello-world/src/activities"
	"time"

	"github.com/google/uuid"
	"go.temporal.io/sdk/workflow"
)

type Repository struct {
	URL            string
	Branch         string
	Path           string
	FileExtensions []string
}
type DocumentsProcessingWorkflowInput struct {
	ID         string
	Repository Repository
}
type DocumentsProcessingWorkflowOutput struct {
	TableName string
}

func DocumentsProcessingWorkflow(ctx workflow.Context, input DocumentsProcessingWorkflowInput) (DocumentsProcessingWorkflowOutput, error) {
	ao := workflow.ActivityOptions{
		StartToCloseTimeout: 1 * time.Minute,
	}
	ctx = workflow.WithActivityOptions(ctx, ao)

	//Create Bucket
	createS3BucketInput := activities.CreateS3BucketInput{Bucket: input.ID}
	err := workflow.ExecuteActivity(ctx, activities.CreateS3Bucket, createS3BucketInput).Get(ctx, nil)
	if err != nil {
		return DocumentsProcessingWorkflowOutput{}, err
	}

	// Collect Documents
	aoMediumTime := workflow.ActivityOptions{
		StartToCloseTimeout: 5 * time.Minute,
	}
	ctx = workflow.WithActivityOptions(ctx, aoMediumTime)

	id := input.ID
	repository := input.Repository
	collectDocumentsInput := activities.CollectDocumentsInput{WorkflowID: id, S3Bucket: id, GitRepoURL: repository.URL, GitRepoBranch: repository.Branch, GitRepoDirectory: repository.Path, FileExtensions: repository.FileExtensions}

	var zipFileName activities.CollectDocumentsOutput
	err = workflow.ExecuteActivity(ctx, activities.CollectDocuments, collectDocumentsInput).Get(ctx, &zipFileName)
	if err != nil {
		return DocumentsProcessingWorkflowOutput{}, err
	}

	// Process Documents
	aoLongTime := workflow.ActivityOptions{
		StartToCloseTimeout: 50 * time.Minute,
	}
	ctx = workflow.WithActivityOptions(ctx, aoLongTime)

	processDocumentsInput := activities.ProcessDocumentsInput{WorkflowID: id, S3Bucket: id, ZipFileName: zipFileName.ZipFileName}

	var tableName activities.ProcessDocumentsOutput
	err = workflow.ExecuteActivity(ctx, activities.ProcessDocuments, processDocumentsInput).Get(ctx, &tableName)
	if err != nil {
		return DocumentsProcessingWorkflowOutput{}, err
	}

	//Delete S3 Object
	ctx = workflow.WithActivityOptions(ctx, ao)
	deleteS3ObjectInput := activities.DeleteS3ObjectInput{Bucket: input.ID, Key: zipFileName.ZipFileName}
	err = workflow.ExecuteActivity(ctx, activities.DeleteS3Object, deleteS3ObjectInput).Get(ctx, nil)
	if err != nil {

		return DocumentsProcessingWorkflowOutput{}, err
	}
	//Delete S3 Bucket
	deleteS3BucketInput := activities.DeleteS3BucketInput{Bucket: input.ID}
	err = workflow.ExecuteActivity(ctx, activities.DeleteS3Bucket, deleteS3BucketInput).Get(ctx, nil)
	if err != nil {

		return DocumentsProcessingWorkflowOutput{}, err
	}

	return DocumentsProcessingWorkflowOutput{TableName: tableName.TableName}, nil
}

type QueryWorkflowInput struct {
	Query          string
	ConversationID string
}

type QueryWorkflowOutput struct {
	ConversationID string
	Response       string
}

func InvokePromptWorkflow(ctx workflow.Context, input QueryWorkflowInput) (QueryWorkflowOutput, error) {
	ao := workflow.ActivityOptions{
		StartToCloseTimeout: 1 * time.Minute,
	}
	ctx = workflow.WithActivityOptions(ctx, ao)

	conversationID := input.ConversationID
	query := input.Query

	//Gets most recent process-document workflow id
	var latestDocumentProcessingID string
	err := workflow.ExecuteActivity(ctx, activities.GetLatestDocumentProcessingId).Get(ctx, &latestDocumentProcessingID)
	if err != nil {
		return QueryWorkflowOutput{}, err
	}

	// Generate a new conversation ID if one isn't provided
	if conversationID == "" {
		conversationID = fmt.Sprintf("conversation-%s", uuid.New().String())
		createS3BucketInput := activities.CreateS3BucketInput{Bucket: conversationID}
		err := workflow.ExecuteActivity(ctx, activities.CreateS3Bucket, createS3BucketInput).Get(ctx, nil)
		if err != nil {
			return QueryWorkflowOutput{}, err
		}
	}

	// Generates conversationFilename
	getRelatedDocumentsInput := activities.GetRelatedDocumentsInput{Query: query, LatestDocumentProcessingId: latestDocumentProcessingID, S3Bucket: conversationID}
	var conversationFilename activities.GetRelatedDocumentsOutput
	err = workflow.ExecuteActivity(ctx, activities.GeneratePrompt, getRelatedDocumentsInput).Get(ctx, &conversationFilename)
	if err != nil {
		return QueryWorkflowOutput{}, err
	}

	// Generates conversationFilename
	invokePromptInput := activities.InvokePromptInput{Query: query, S3Bucket: conversationID, ConversationFilename: conversationFilename.ConversationFilename}
	var response activities.InvokePromptOutput
	err = workflow.ExecuteActivity(ctx, activities.InvokePrompt, invokePromptInput).Get(ctx, &response)
	if err != nil {
		return QueryWorkflowOutput{}, err
	}

	return QueryWorkflowOutput{
		ConversationID: conversationID,
		Response:       response.Response,
	}, nil
}
